components/eamxx/src/physics/p3/impl/p3_init_impl.hpp

```diff
@@ -3,35 +3,327 @@
 
 #include "p3_functions.hpp" // for ETI only but harmless for GPU
 
-extern "C" {
-  void micro_p3_utils_init_c(scream::Real Cpair, scream::Real Rair, scream::Real RH2O, scream::Real RHO_H2O,
-                 scream::Real MWH2O, scream::Real MWdry, scream::Real gravit, scream::Real LatVap, scream::Real LatIce,
-                 scream::Real CpLiq, scream::Real Tmelt, scream::Real Pi, bool masterproc);
-  void p3_init_c(const char** lookup_file_dir, int* info, const bool& write_tables);
-}
+#include "ekat/util/ekat_file_utils.hpp"
+
+#include <fstream>
 
 namespace scream {
 namespace p3 {
 
+namespace {
+
+template <typename S, typename IceT, typename CollT>
+void read_ice_lookup_tables(const bool masterproc, const char* p3_lookup_base, const char* p3_version, IceT& ice_table_vals, CollT& collect_table_vals, int densize, int rimsize, int isize, int rcollsize)
+{
+  using DeviceIcetable = typename IceT::non_const_type;
+  using DeviceColtable = typename CollT::non_const_type;
+
+  const auto ice_table_vals_d     = DeviceIcetable("ice_table_vals");
+  const auto collect_table_vals_d = DeviceColtable("collect_table_vals");
+
+  const auto ice_table_vals_h    = Kokkos::create_mirror_view(ice_table_vals_d);
+  const auto collect_table_vals_h = Kokkos::create_mirror_view(collect_table_vals_d);
+
+  //
+  // read in ice microphysics table into host views. We always read these as doubles.
+  //
+
+  std::string filename = std::string(p3_lookup_base) + std::string(p3_version);
+
+  if (masterproc) {
+    std::cout << "Reading ice lookup tables in file: " << filename << std::endl;
+  }
+
+  std::ifstream in(filename);
+
+  // read header
+  std::string version, version_val;
+  in >> version >> version_val;
+  EKAT_REQUIRE_MSG(version == "VERSION", "Bad " << filename << ", expected VERSION X.Y.Z header");
+  EKAT_REQUIRE_MSG(version_val == p3_version, "Bad " << filename << ", expected version " << p3_version << ", but got " << version_val);
+
+  // read tables
+  double dum_s; int dum_i; // dum_s needs to be double to stream correctly
```

Comment:
Are these just dummy vars to swallow entries that are in the file but not used? Do we need to keep these unused entries in the file or can we just prune them, and keep only what is used?

Reply:
It looks like `dum_s` is used but not `dum_i`. We probably don't need the unused entries, but I don't really know the history of this file.